File: skills/model-validation-and-uncertainty/scripts/extrapolation_risk.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Usage: Rscript extrapolation_risk.R <training_raster_stack.tif> <projection_raster_stack.tif> <output_dir>
#
# Arguments:
#   training_raster_stack.tif   : Multi-band GeoTIFF used for model calibration
#   projection_raster_stack.tif : Multi-band GeoTIFF for the projection area/period
#   output_dir                  : Directory for outputs (created if absent)
#
# Outputs:
#   mop_layer.tif              — MOP raster (0 = strict extrapolation, 1 = fully within range)
#   mess_layer.tif             — MESS raster (negative = novel environment)
#   extrapolation_summary.csv  — Summary statistics (% area per threshold)
#   extrapolation_plots.png    — Side-by-side MOP and MESS maps

# ── Inline logger ─────────────────────────────────────────────────────────────
SKILL_NAME <- "model-validation-and-uncertainty"
.log_ts  <- function() format(Sys.time(), "[%Y-%m-%d %H:%M:%S]")
log_info <- function(...) message(.log_ts(), " [INFO]  ", sprintf(...))
log_warn <- function(...) message(.log_ts(), " [WARN]  ", sprintf(...))
log_error<- function(...) message(.log_ts(), " [ERROR] ", sprintf(...))
log_step <- function(n, d) log_info("-- STEP %d: %s", n, d)
log_decision <- function(v, val, why) log_info("DECISION | %s = %s | %s", v, val, why)
dir.create("logs", recursive=TRUE, showWarnings=FALSE)

suppressPackageStartupMessages(library(terra))
suppressPackageStartupMessages(library(dismo))
suppressPackageStartupMessages(library(ggplot2))

# ── 1. Parse arguments ──────────────────────────────────────────────────────
log_step(1, "Parse arguments and validate inputs")
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 3) {
  log_warn("Fewer than 3 arguments provided. Using default paths for testing.")
  train_path  <- "data/predictors/env_train.tif"
  proj_path   <- "data/predictors/env_proj.tif"
  output_dir  <- "output/extrapolation"
} else {
  train_path  <- args[1]
  proj_path   <- args[2]
  output_dir  <- args[3]
}

log_decision("train_path", train_path, "raster stack used for model calibration")
log_decision("proj_path",  proj_path,  "raster stack for the projection area/period")

if (!file.exists(train_path)) {
  log_error(
    "Falha em validate inputs: raster de treinamento nao encontrado: %s\nCausa provavel: caminho incorreto ou arquivo GeoTIFF nao gerado\nVerifique: o argumento training_raster_stack.tif e o diretorio de trabalho\nSkill anterior: species-distribution-modelling",
    train_path
  )
  stop("Training raster not found.")
}
if (!file.exists(proj_path)) {
  log_error(
    "Falha em validate inputs: raster de projecao nao encontrado: %s\nCausa provavel: caminho incorreto ou arquivo GeoTIFF nao gerado\nVerifique: o argumento projection_raster_stack.tif e o diretorio de trabalho\nSkill anterior: species-distribution-modelling",
    proj_path
  )
  stop("Projection raster not found.")
}

# ── 2. Create output directory ───────────────────────────────────────────────
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

# ── 3. Load raster stacks ────────────────────────────────────────────────────
log_step(2, "Load raster stacks")
tryCatch({
  log_info("Loading training stack: %s", train_path)
  train_stack <- rast(train_path)

  log_info("Loading projection stack: %s", proj_path)
  proj_stack  <- rast(proj_path)
}, error = function(e) {
  log_error(
    "Falha em load rasters: %s\nCausa provavel: arquivo GeoTIFF corrompido ou formato nao suportado\nVerifique: integridade dos arquivos TIF com gdalinfo\nSkill anterior: species-distribution-modelling",
    conditionMessage(e)
  )
  stop(e)
})

# Validate that both stacks have the same layers
if (!setequal(names(train_stack), names(proj_stack))) {
  mismatched <- setdiff(names(train_stack), names(proj_stack))
  log_error(
    "Falha em validate layers: nomes de camadas divergem entre stacks de treinamento e projecao.\nCamadas ausentes na projecao: %s\nCausa provavel: stacks gerados com variaveis diferentes\nVerifique: que ambos os TIFs tem as mesmas bandas nomeadas\nSkill anterior: species-distribution-modelling",
    paste(mismatched, collapse = ", ")
  )
  stop("Layer name mismatch between training and projection stacks.\n  Missing in projection: ",
       paste(mismatched, collapse = ", "))
}

# Reorder projection layers to match training layer order
proj_stack <- proj_stack[[names(train_stack)]]
n_vars <- nlyr(train_stack)
log_info("Variables (%d): %s", n_vars, paste(names(train_stack), collapse = ", "))

# ── 4. Extract calibration reference values ──────────────────────────────────
log_step(3, "Extract calibration reference values")
tryCatch({
  cal_vals <- as.data.frame(train_stack, na.rm = TRUE)
  log_info("Calibration pixels extracted: %d", nrow(cal_vals))

  if (nrow(cal_vals) < 100) {
    log_warn("Only %d non-NA calibration pixels. MOP estimates may be unstable.", nrow(cal_vals))
  }

  # Scale calibration values for MOP distance computation
  cal_center <- colMeans(cal_vals, na.rm = TRUE)
  cal_sd     <- apply(cal_vals, 2, sd, na.rm = TRUE)

  zero_sd_vars <- names(cal_sd)[cal_sd == 0]
  if (length(zero_sd_vars) > 0) {
    log_warn("Variables with zero variance (will be set to sd=1): %s", paste(zero_sd_vars, collapse = ", "))
  }

  # Replace zero sd with 1 to avoid division by zero
  cal_sd[cal_sd == 0] <- 1

  cal_scaled <- scale(cal_vals, center = cal_center, scale = cal_sd)
  n_cal      <- nrow(cal_scaled)
  log_decision("mop_scaling", "z-score using calibration mean/sd", "ensures all variables contribute equally to Euclidean distance")
}, error = function(e) {
  log_error(
    "Falha em extract calibration values: %s\nCausa provavel: raster de treinamento com todos os pixels NA\nVerifique: mascara e extent do raster de treinamento\nSkill anterior: species-distribution-modelling",
    conditionMessage(e)
  )
  stop(e)
})

# ── 5. Compute MOP (Mobility-Oriented Parity) ────────────────────────────────
# Reference: Owens et al. 2013. Ecol. Model. 263:10-18.
# DOI: 10.1016/j.ecolmodel.2013.04.011
#
# For each projection pixel, MOP = proportion of calibration points that are
# "closer" (in standardised Euclidean space) than the projection pixel.
# MOP = 0 means the pixel is beyond ALL calibration points — strict extrapolation.

log_step(4, "Compute MOP layer (Owens et al. 2013)")
log_decision("mop_percentile", "10th percentile of pixel-to-calibration distances", "standard implementation following Owens et al. 2013")
tryCatch({
  log_info("Computing MOP layer (this may take a few minutes)...")

  # Compute the centroid distance of each calibration point
  cal_centroid_dist <- sqrt(rowSums(cal_scaled^2))

  # Apply MOP computation pixel by pixel using terra::app
  proj_vals <- as.data.frame(proj_stack, na.rm = FALSE, xy = TRUE)
  xy_cols   <- c("x", "y")
  env_cols  <- setdiff(names(proj_vals), xy_cols)

  mop_compute <- function(px_env) {
    if (any(is.na(px_env))) return(NA_real_)

    # Scale projection pixel using calibration parameters
    px_scaled <- (as.numeric(px_env) - cal_center) / cal_sd

    # Euclidean distance from this pixel to every calibration point
    d_px_to_cal <- sqrt(rowSums(sweep(cal_scaled, 2, px_scaled, "-")^2))

    # MOP = proportion of calibration points whose centroid distance
    # is less than the 10th percentile of distances from this pixel
    ref_dist <- quantile(d_px_to_cal, 0.1)
    sum(cal_centroid_dist < ref_dist) / n_cal
  }

  mop_vals <- apply(proj_vals[, env_cols], 1, mop_compute)

  # Reconstruct as SpatRaster
  mop_rast        <- rast(proj_stack[[1]])
  values(mop_rast) <- NA
  # Map back to all pixels (including those with NA that were skipped)
  full_vals                                <- rep(NA_real_, ncell(mop_rast))
  non_na_idx                               <- which(!is.na(values(proj_stack[[1]])))
  full_vals[non_na_idx[seq_along(mop_vals)]] <- mop_vals
  values(mop_rast) <- full_vals
  names(mop_rast)  <- "MOP"

  # Save MOP raster
  mop_path <- file.path(output_dir, "mop_layer.tif")
  writeRaster(mop_rast, mop_path, overwrite = TRUE)
  log_info("Saved: %s", mop_path)
}, error = function(e) {
  log_error(
    "Falha em MOP computation: %s\nCausa provavel: memoria insuficiente para rasters grandes ou valores NA inesperados\nVerifique: tamanho do raster de projecao e memoria disponivel\nSkill anterior: model-validation-and-uncertainty (calibration extraction)",
    conditionMessage(e)
  )
  stop(e)
})

# ── 6. Compute MESS (Multivariate Environmental Similarity Surfaces) ─────────
# Reference: Elith et al. 2010. Meth. Ecol. Evol. 1:330-342.
# DOI: 10.1111/j.2041-210X.2010.00036.x
#
# MESS < 0 indicates novel environment relative to calibration reference set.

log_step(5, "Compute MESS layer (Elith et al. 2010)")
tryCatch({
  log_info("Computing MESS layer...")

  # dismo::mess requires a RasterStack (terra → raster conversion for compatibility)
  suppressPackageStartupMessages(library(raster))
  proj_raster <- raster::stack(proj_stack)
  train_df    <- cal_vals  # reference points

  mess_result <- dismo::mess(proj_raster, train_df, full = FALSE)

  # Convert back to terra SpatRaster
  mess_rast  <- rast(mess_result)
  names(mess_rast) <- "MESS"

  # Save MESS raster
  mess_path <- file.path(output_dir, "mess_layer.tif")
  writeRaster(mess_rast, mess_path, overwrite = TRUE)
  log_info("Saved: %s", mess_path)
}, error = function(e) {
  log_error(
    "Falha em MESS computation: %s\nCausa provavel: incompatibilidade entre pacotes terra/raster ou raster sem CRS\nVerifique: versoes de terra e dismo, e que os rasters tem CRS definido\nSkill anterior: model-validation-and-uncertainty (calibration extraction)",
    conditionMessage(e)
  )
  stop(e)
})

# ── 7. Compute summary statistics ────────────────────────────────────────────
log_step(6, "Compute extrapolation summary statistics")
tryCatch({
  mop_v  <- values(mop_rast,  na.rm = TRUE)
  mess_v <- values(mess_rast, na.rm = TRUE)
  n_proj <- length(mop_v)

  pct_mop_zero <- round(100 * sum(mop_v == 0,     na.rm = TRUE) / n_proj, 2)
  pct_mop_025  <- round(100 * sum(mop_v < 0.25,   na.rm = TRUE) / n_proj, 2)
  pct_mop_050  <- round(100 * sum(mop_v < 0.50,   na.rm = TRUE) / n_proj, 2)
  pct_mess_neg <- round(100 * sum(mess_v < 0,      na.rm = TRUE) / length(mess_v[!is.na(mess_v)]), 2)

  summary_df <- data.frame(
    metric = c("pct_area_MOP_zero",
               "pct_area_MOP_lt_0.25",
               "pct_area_MOP_lt_0.50",
               "pct_area_MESS_negative"),
    value  = c(pct_mop_zero, pct_mop_025, pct_mop_050, pct_mess_neg),
    interpretation = c(
      "Strict extrapolation (MOP = 0)",
      "High novelty (MOP < 0.25)",
      "Moderate-high novelty (MOP < 0.50)",
      "Novel environment in MESS (MESS < 0)"
    )
  )

  csv_path <- file.path(output_dir, "extrapolation_summary.csv")
  write.csv(summary_df, csv_path, row.names = FALSE)
  log_info("Saved: %s", csv_path)
}, error = function(e) {
  log_error(
    "Falha em summary statistics: %s\nCausa provavel: rasters MOP ou MESS invalidos\nVerifique: etapas anteriores para mensagens de erro\nSkill anterior: model-validation-and-uncertainty (MOP/MESS computation)",
    conditionMessage(e)
  )
  stop(e)
})

# ── 8. Automatic warning if extrapolation is severe ──────────────────────────
if (pct_mop_025 > 30) {
  log_warn(
    "EXTRAPOLATION WARNING: %.1f%% of the projection area has MOP < 0.25 (high novelty relative to calibration). Predictions in these areas should be treated with extreme caution. Recommendation: mask MOP < 0.25 pixels in publication figures and add explicit caveats in the methods section.",
    pct_mop_025
  )
}

if (pct_mop_zero > 10) {
  log_warn(
    "STRICT EXTRAPOLATION WARNING: %.1f%% of the projection area has MOP = 0 (model extrapolates beyond all calibration data). These pixels MUST be masked in publication figures.",
    pct_mop_zero
  )
}

# ── 9. Side-by-side diagnostic plots ─────────────────────────────────────────
log_step(7, "Generate extrapolation diagnostic plots")
tryCatch({
  png(file.path(output_dir, "extrapolation_plots.png"),
      width = 1600, height = 700, res = 150)
  par(mfrow = c(1, 2), mar = c(4, 4, 3, 5))

  # MOP map
  plot(mop_rast, main = "MOP (0 = strict extrapolation)",
       col = rev(terrain.colors(100)),
       legend = TRUE, axes = FALSE)
  mtext(paste0("MOP = 0: ", pct_mop_zero, "% | MOP < 0.25: ", pct_mop_025, "%"),
        side = 1, cex = 0.8)

  # MESS map (diverging palette: red = novel, blue = similar)
  mess_cols <- colorRampPalette(c("red", "white", "steelblue"))(100)
  plot(mess_rast, main = "MESS (negative = novel environment)",
       col = mess_cols,
       legend = TRUE, axes = FALSE)
  mtext(paste0("MESS < 0: ", pct_mess_neg, "%"),
        side = 1, cex = 0.8)

  dev.off()
  log_info("Saved: %s", file.path(output_dir, "extrapolation_plots.png"))
}, error = function(e) {
  log_error(
    "Falha em diagnostic plots: %s\nCausa provavel: dispositivo grafico nao disponivel ou rasters invalidos\nVerifique: disponibilidade de X11/display e integridade dos rasters\nSkill anterior: model-validation-and-uncertainty (MOP/MESS computation)",
    conditionMessage(e)
  )
  stop(e)
})

# ── 10. Final summary ─────────────────────────────────────────────────────────
log_info("========== EXTRAPOLATION SUMMARY ==========")
log_info("%% area MOP = 0    (strict extrapolation): %.2f%%", pct_mop_zero)
log_info("%% area MOP < 0.25 (high novelty)        : %.2f%%", pct_mop_025)
log_info("%% area MOP < 0.50 (moderate novelty)    : %.2f%%", pct_mop_050)
log_info("%% area MESS < 0   (novel environment)   : %.2f%%", pct_mess_neg)
log_info("===========================================")

File: skills/model-validation-and-uncertainty/scripts/extrapolation_risk.py
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Usage: python extrapolation_risk.py <training_raster_stack.tif> <projection_raster_stack.tif> <output_dir>
#
# Arguments:
#   training_raster_stack.tif   : Multi-band GeoTIFF used for model calibration
#   projection_raster_stack.tif : Multi-band GeoTIFF for the projection area/period
#   output_dir                  : Directory for outputs (created if absent)
#
# Outputs:
#   mop_layer.tif              - MOP raster (0 = strict extrapolation, 1 = fully within range)
#   mess_layer.tif             - MESS raster (negative = novel environment)
#   extrapolation_summary.csv  - Summary statistics (% area per threshold)
#   extrapolation_plots.png    - Side-by-side MOP and MESS maps

import sys
import os
import logging
import datetime
import numpy as np
import pandas as pd
from scipy.spatial.distance import cdist
import rasterio
from rasterio.transform import from_bounds
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors

# -- Inline logger ------------------------------------------------------------
SKILL_NAME = "model-validation-and-uncertainty"
logging.basicConfig(
    level=logging.INFO,
    format="[%(asctime)s] [%(levelname)s] %(message)s",
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger(SKILL_NAME)
os.makedirs("logs", exist_ok=True)


def log_decision(variable: str, value: str, reason: str) -> None:
    """Log a methodological decision."""
    logger.info("DECISION | %s = %s | %s", variable, value, reason)



def main():
    # -- 1. Parse arguments -------------------------------------------------------
    logger.info("-- STEP 1: Parse arguments and validate inputs")
    if len(sys.argv) < 4:
        logger.warning("Fewer than 3 arguments provided. Using default paths for testing.")
        train_path = "data/predictors/env_train.tif"
        proj_path = "data/predictors/env_proj.tif"
        output_dir = "output/extrapolation"
    else:
        train_path = sys.argv[1]
        proj_path = sys.argv[2]
        output_dir = sys.argv[3]

    log_decision("train_path", train_path, "raster stack used for model calibration")
    log_decision("proj_path", proj_path, "raster stack for the projection area/period")

    if not os.path.isfile(train_path):
        logger.error(
            "Failed to validate inputs: training raster not found: %s\n"
            "Probable cause: incorrect path or GeoTIFF file not generated\n"
            "Check: the training_raster_stack.tif argument and the working directory\n"
            "Previous skill: species-distribution-modelling",
            train_path,
        )
        sys.exit(1)

    if not os.path.isfile(proj_path):
        logger.error(
            "Failed in validate inputs: projection raster not found: %s\n"
            "Probable cause: incorrect path or GeoTIFF file not yet generated\n"
            "Check: the projection_raster_stack.tif argument and working directory\n"
            "Previous skill: species-distribution-modelling",
            proj_path,
        )
        sys.exit(1)

    # -- 2. Create output directory ------------------------------------------------
    os.makedirs(output_dir, exist_ok=True)

    # -- 3. Load raster stacks ----------------------------------------------------
    logger.info("-- STEP 2: Load raster stacks")
    try:
        logger.info("Loading training stack: %s", train_path)
        train_ds = rasterio.open(train_path)
        train_bands = train_ds.read()  # shape: (n_bands, rows, cols)
        train_names = [
            train_ds.descriptions[i] if train_ds.descriptions[i] else f"band_{i+1}"
            for i in range(train_ds.count)
        ]

        logger.info("Loading projection stack: %s", proj_path)
        proj_ds = rasterio.open(proj_path)
        proj_bands = proj_ds.read()
        proj_names = [
            proj_ds.descriptions[i] if proj_ds.descriptions[i] else f"band_{i+1}"
            for i in range(proj_ds.count)
        ]
    except Exception as e:
        logger.error(
            "Failed in load rasters: %s\n"
            "Probable cause: corrupted GeoTIFF file or format not supported\n"
            "Check: TIF file integrity using gdalinfo\n"
            "Previous skill: species-distribution-modelling",
            e,
        )
        sys.exit(1)

    # Validate that both stacks have the same layers
    if set(train_names) != set(proj_names):
        missing = set(train_names) - set(proj_names)
        logger.error(
            "Failed in validate layers: layer names differ between training and projection stacks.\n"
            "Layers missing in projection: %s\n"
            "Probable cause: stacks generated with different variables\n"
            "Check: that both TIFs have the same named bands\n"
            "Previous skill: species-distribution-modelling",
            ", ".join(missing),
        )
        sys.exit(1)

    # Reorder projection layers to match training layer order
    proj_order = [proj_names.index(n) for n in train_names]
    proj_bands = proj_bands[proj_order]
    n_vars = train_ds.count
    logger.info("Variables (%d): %s", n_vars, ", ".join(train_names))

    # -- 4. Extract calibration reference values -----------------------------------
    logger.info("-- STEP 3: Extract calibration reference values")
    try:
        # Reshape to (n_pixels, n_bands) and remove NA rows
        train_flat = train_bands.reshape(n_vars, -1).T  # (n_pixels, n_vars)
        train_nodata = train_ds.nodata
        mask_train = np.all(np.isfinite(train_flat), axis=1)
        if train_nodata is not None:
            mask_train &= np.all(train_flat != train_nodata, axis=1)
        cal_vals = train_flat[mask_train]
        logger.info("Calibration pixels extracted: %d", cal_vals.shape[0])

        if cal_vals.shape[0] < 100:
            logger.warning(
                "Only %d non-NA calibration pixels. MOP estimates may be unstable.",
                cal_vals.shape[0],
            )

        # Scale calibration values for MOP distance computation
        cal_center = np.nanmean(cal_vals, axis=0)
        cal_sd = np.nanstd(cal_vals, axis=0, ddof=1)

        zero_sd_vars = [train_names[i] for i, s in enumerate(cal_sd) if s == 0]
        if zero_sd_vars:
            logger.warning(
                "Variables with zero variance (will be set to sd=1): %s",
                ", ".join(zero_sd_vars),
            )

        cal_sd[cal_sd == 0] = 1.0
        cal_scaled = (cal_vals - cal_center) / cal_sd
        n_cal = cal_scaled.shape[0]
        log_decision(
            "mop_scaling",
            "z-score using calibration mean/sd",
            "ensures all variables contribute equally to Euclidean distance",
        )
    except Exception as e:
        logger.error(
            "Failed in extract calibration values: %s\n"
            "Probable cause: training raster with all NA pixels\n"
            "Check: training raster mask and extent\n"
            "Previous skill: species-distribution-modelling",
            e,
        )
        sys.exit(1)

    # -- 5. Compute MOP (Mobility-Oriented Parity) --------------------------------
    # Reference: Owens et al. 2013. Ecol. Model. 263:10-18.
    # DOI: 10.1016/j.ecolmodel.2013.04.011
    #
    # For each projection pixel, MOP = proportion of calibration points that are
    # "closer" (in standardised Euclidean space) than the projection pixel.
    # MOP = 0 means the pixel is beyond ALL calibration points -- strict extrapolation.

    logger.info("-- STEP 4: Compute MOP layer (Owens et al. 2013)")
    log_decision(
        "mop_percentile",
        "10th percentile of pixel-to-calibration distances",
        "standard implementation following Owens et al. 2013",
    )
    try:
        logger.info("Computing MOP layer (this may take a few minutes)...")

        # Compute pairwise distances among calibration points (reference distribution)
        cal_pairwise_dist = cdist(cal_scaled, cal_scaled, metric="euclidean")
        # 10th percentile of non-zero distances per calibration point
        cal_ref_dist = np.array(
            [
                np.percentile(row[row > 0], 10) if np.any(row > 0) else 0.0
                for row in cal_pairwise_dist
            ]
        )

        # Projection pixels: reshape and scale
        proj_flat = proj_bands.reshape(n_vars, -1).T  # (n_pixels, n_vars)
        proj_nodata = proj_ds.nodata
        mask_proj = np.all(np.isfinite(proj_flat), axis=1)
        if proj_nodata is not None:
            mask_proj &= np.all(proj_flat != proj_nodata, axis=1)

        proj_valid = proj_flat[mask_proj]
        proj_scaled = (proj_valid - cal_center) / cal_sd

        # Compute MOP for each valid projection pixel
        # Process in chunks to manage memory
        chunk_size = 5000
        mop_valid = np.empty(proj_scaled.shape[0], dtype=np.float64)

        for start in range(0, proj_scaled.shape[0], chunk_size):
            end = min(start + chunk_size, proj_scaled.shape[0])
            chunk = proj_scaled[start:end]
            # Distance from each chunk pixel to every calibration point
            d_chunk = cdist(chunk, cal_scaled, metric="euclidean")
            for i in range(d_chunk.shape[0]):
                px_ref_dist = np.percentile(d_chunk[i], 10)
                mop_valid[start + i] = np.sum(cal_ref_dist >= px_ref_dist) / n_cal

        # Reconstruct full raster
        mop_full = np.full(proj_flat.shape[0], np.nan, dtype=np.float64)
        mop_full[mask_proj] = mop_valid
        mop_rast = mop_full.reshape(proj_bands.shape[1], proj_bands.shape[2])

        # Save MOP raster
        mop_path = os.path.join(output_dir, "mop_layer.tif")
        profile = proj_ds.profile.copy()
        profile.update(count=1, dtype="float64", nodata=np.nan)
        with rasterio.open(mop_path, "w", **profile) as dst:
            dst.write(mop_rast, 1)
            dst.set_band_description(1, "MOP")
        logger.info("Saved: %s", mop_path)

    except Exception as e:
        logger.error(
            "Failed in MOP computation: %s\n"
            "Probable cause: insufficient memory for large rasters or unexpected NA values\n"
            "Check: projection raster size and available memory\n"
            "Previous skill: model-validation-and-uncertainty (calibration extraction)",
            e,
        )
        sys.exit(1)

    # -- 6. Compute MESS (Multivariate Environmental Similarity Surfaces) ----------
    # Reference: Elith et al. 2010. Meth. Ecol. Evol. 1:330-342.
    # DOI: 10.1111/j.2041-210X.2010.00036.x
    #
    # MESS < 0 indicates novel environment relative to calibration reference set.

    logger.info("-- STEP 5: Compute MESS layer (Elith et al. 2010)")
    try:
        logger.info("Computing MESS layer...")

        def compute_mess_pixel(px_env: np.ndarray, cal_data: np.ndarray) -> float:
            """Compute MESS for a single pixel across all variables.

            For each variable, similarity is:
              - If px < min(cal): 100 * (px - min) / (max - min)   [negative]
              - If px > max(cal): 100 * (max - px) / (max - min)   [negative]
              - Otherwise: min(f, 100-f) where f = 100 * (sum(cal <= px) / n)
            MESS = min across all variables.
            """
            n = cal_data.shape[0]
            sims = np.empty(cal_data.shape[1])
            for j in range(cal_data.shape[1]):
                col = cal_data[:, j]
                mn, mx = col.min(), col.max()
                rng = mx - mn
                if rng == 0:
                    sims[j] = 0.0
                    continue
                v = px_env[j]
                if v < mn:
                    sims[j] = 100.0 * (v - mn) / rng
                elif v > mx:
                    sims[j] = 100.0 * (mx - v) / rng
                else:
                    f = 100.0 * np.sum(col <= v) / n
                    sims[j] = min(f, 100.0 - f)
            return np.min(sims)

        # Compute MESS for all valid projection pixels
        mess_valid = np.array(
            [compute_mess_pixel(proj_valid[i], cal_vals) for i in range(proj_valid.shape[0])]
        )

        mess_full = np.full(proj_flat.shape[0], np.nan, dtype=np.float64)
        mess_full[mask_proj] = mess_valid
        mess_rast = mess_full.reshape(proj_bands.shape[1], proj_bands.shape[2])

        # Save MESS raster
        mess_path = os.path.join(output_dir, "mess_layer.tif")
        profile_mess = proj_ds.profile.copy()
        profile_mess.update(count=1, dtype="float64", nodata=np.nan)
        with rasterio.open(mess_path, "w", **profile_mess) as dst:
            dst.write(mess_rast, 1)
            dst.set_band_description(1, "MESS")
        logger.info("Saved: %s", mess_path)

    except Exception as e:
        logger.error(
            "Failed in MESS computation: %s\n"
            "Probable cause: incompatibility between rasterio or raster without CRS\n"
            "Check: rasterio version and that rasters have CRS defined\n"
            "Previous skill: model-validation-and-uncertainty (calibration extraction)",
            e,
        )
        sys.exit(1)

    # -- 7. Compute summary statistics --------------------------------------------
    logger.info("-- STEP 6: Compute extrapolation summary statistics")
    try:
        mop_v = mop_full[~np.isnan(mop_full)]
        mess_v = mess_full[~np.isnan(mess_full)]
        n_proj = len(mop_v)

        pct_mop_zero = round(100.0 * np.sum(mop_v == 0) / n_proj, 2)
        pct_mop_025 = round(100.0 * np.sum(mop_v < 0.25) / n_proj, 2)
        pct_mop_050 = round(100.0 * np.sum(mop_v < 0.50) / n_proj, 2)
        pct_mess_neg = round(100.0 * np.sum(mess_v < 0) / len(mess_v), 2)

        summary_df = pd.DataFrame(
            {
                "metric": [
                    "pct_area_MOP_zero",
                    "pct_area_MOP_lt_0.25",
                    "pct_area_MOP_lt_0.50",
                    "pct_area_MESS_negative",
                ],
                "value": [pct_mop_zero, pct_mop_025, pct_mop_050, pct_mess_neg],
                "interpretation": [
                    "Strict extrapolation (MOP = 0)",
                    "High novelty (MOP < 0.25)",
                    "Moderate-high novelty (MOP < 0.50)",
                    "Novel environment in MESS (MESS < 0)",
                ],
            }
        )

        csv_path = os.path.join(output_dir, "extrapolation_summary.csv")
        summary_df.to_csv(csv_path, index=False)
        logger.info("Saved: %s", csv_path)

    except Exception as e:
        logger.error(
            "Failed in summary statistics: %s\n"
            "Probable cause: invalid MOP or MESS rasters\n"
            "Check: previous steps for error messages\n"
            "Previous skill: model-validation-and-uncertainty (MOP/MESS computation)",
            e,
        )
        sys.exit(1)

    # -- 8. Automatic warning if extrapolation is severe ---------------------------
    if pct_mop_025 > 30:
        logger.warning(
            "EXTRAPOLATION WARNING: %.1f%% of the projection area has MOP < 0.25 "
            "(high novelty relative to calibration). Predictions in these areas should "
            "be treated with extreme caution. Recommendation: mask MOP < 0.25 pixels in "
            "publication figures and add explicit caveats in the methods section.",
            pct_mop_025,
        )

    if pct_mop_zero > 10:
        logger.warning(
            "STRICT EXTRAPOLATION WARNING: %.1f%% of the projection area has MOP = 0 "
            "(model extrapolates beyond all calibration data). These pixels MUST be "
            "masked in publication figures.",
            pct_mop_zero,
        )

    # -- 9. Side-by-side diagnostic plots -----------------------------------------
    logger.info("-- STEP 7: Generate extrapolation diagnostic plots")
    try:
        fig, axes = plt.subplots(1, 2, figsize=(16, 7), dpi=150)

        # MOP map
        cmap_mop = plt.cm.terrain_r
        im_mop = axes[0].imshow(mop_rast, cmap=cmap_mop, vmin=0, vmax=1)
        axes[0].set_title("MOP (0 = strict extrapolation)")
        axes[0].set_xlabel(f"MOP = 0: {pct_mop_zero}% | MOP < 0.25: {pct_mop_025}%")
        axes[0].set_xticks([])
        axes[0].set_yticks([])
        fig.colorbar(im_mop, ax=axes[0], fraction=0.046, pad=0.04)

        # MESS map (diverging palette: red = novel, blue = similar)
        cmap_mess = mcolors.LinearSegmentedColormap.from_list(
            "mess_diverge", ["red", "white", "steelblue"]
        )
        vabs = max(abs(np.nanmin(mess_rast)), abs(np.nanmax(mess_rast)))
        im_mess = axes[1].imshow(mess_rast, cmap=cmap_mess, vmin=-vabs, vmax=vabs)
        axes[1].set_title("MESS (negative = novel environment)")
        axes[1].set_xlabel(f"MESS < 0: {pct_mess_neg}%")
        axes[1].set_xticks([])
        axes[1].set_yticks([])
        fig.colorbar(im_mess, ax=axes[1], fraction=0.046, pad=0.04)

        plt.tight_layout()
        plot_path = os.path.join(output_dir, "extrapolation_plots.png")
        fig.savefig(plot_path)
        plt.close(fig)
        logger.info("Saved: %s", plot_path)

    except Exception as e:
        logger.error(
            "Failed in diagnostic plots: %s\n"
            "Probable cause: graphics backend not available or invalid rasters\n"
            "Check: matplotlib backend availability and raster integrity\n"
            "Previous skill: model-validation-and-uncertainty (MOP/MESS computation)",
            e,
        )
        sys.exit(1)

    # -- 10. Final summary --------------------------------------------------------
    logger.info("========== EXTRAPOLATION SUMMARY ==========")
    logger.info("%% area MOP = 0    (strict extrapolation): %.2f%%", pct_mop_zero)
    logger.info("%% area MOP < 0.25 (high novelty)        : %.2f%%", pct_mop_025)
    logger.info("%% area MOP < 0.50 (moderate novelty)    : %.2f%%", pct_mop_050)
    logger.info("%% area MESS < 0   (novel environment)   : %.2f%%", pct_mess_neg)
    logger.info("===========================================")

    # Clean up
    train_ds.close()
    proj_ds.close()


if __name__ == "__main__":
    main()
